plaidml2/edsl/edsl_test.py

```diff
@@ -99,9 +171,465 @@ def arg_max(I):
+    A.bind_dims(I, K)
+    B.bind_dims(K, J)
+    C = TensorOutput(I, J)
+    C[i, j] += A[i, k] * B[k, j]
+    return C
+
+
+def global_min(I):
+    i, j, k = TensorIndexes(3)
+    Neg = -I
+    O_Neg = TensorOutput()
+    O_Neg[()] >= Neg[i, j, k]
+    O = -O_Neg
+    return O
+
+
+def avg(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[y] += I[x, y]
+    return Sum / X
+
+
+def avg_stages(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[()] += I[x, y]
+    PartialMean = Sum / X
+    return PartialMean / Y
+
+
+def avg_merge(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[()] += I[x, y]
+    return Sum / (X * Y)
+
+
+def skip(I):
+    M, N = TensorDims(2)
+    i, j = TensorIndexes(2)
+    I.bind_dims(M, N)
+    O = TensorOutput(N)
+    O[2 * i] += I[2 * i, j]
+    return O
+
+
+def csum(I):
+    N = TensorDim()
+    i, k = TensorIndexes(2)
+    I.bind_dims(N)
+    O = TensorOutput(N)
+    O[i] += I[k]
+    O.add_constraint(i - k < N)
+    return O
+
+
 class TestEdsl(unittest.TestCase):
     maxDiff = None
 
+    def test_sum_over_axis(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 784]))
+        O = sum_over_axis(I)
+        program = Program('sum_over_axis', [O])
+        if USE_MLIR():
+            self.assertMultiLineEqual(
+                str(program), ('''
+                
+#map0 = (d0, d1) -> (d0)
+#map1 = (d0, d1) -> (d1, d0)
+
+
+!fp32 = type tensor<!eltwise.fp32>
+module {
+  func @sum_over_axis(%arg0: tensor<1x784x!eltwise.fp32>) -> tensor<784x!eltwise.fp32> {
+    %cst = "eltwise.sconst"() {value = 0.000000e+00 : f64} : () -> !fp32
+    %0 = tile.cion add, none, %cst, %arg0 {sink = #map0, srcs = [#map1]} : !fp32, tensor<1x784x!eltwise.fp32> -> tensor<784x!eltwise.fp32>
+    return %0 : tensor<784x!eltwise.fp32>
+  }
+}
+''').lstrip())
+        else:
+            self.assertMultiLineEqual(
+                str(program), '''function (
+  _X0[_X0_0, _X0_1]
+) -> (
+  _X1
+) {
+  _X1[x1 : 784] = +(_X0[x0, x1]);
+}
+''')
+
+    def test_max_over_axis(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 784]))
+        O = max_over_axis(I)
+        program = Program('max_over_axis', [O])
+        if USE_MLIR():
+            self.assertMultiLineEqual(
+                str(program), ('''
+                
+#map0 = (d0, d1) -> (d0)
+#map1 = (d0, d1) -> (d1, d0)
+
+
+!fp32 = type tensor<!eltwise.fp32>
+module {
+  func @max_over_axis(%arg0: tensor<1x784x!eltwise.fp32>) -> tensor<784x!eltwise.fp32> {
+    %cst = "eltwise.sconst"() {value = 0.000000e+00 : f64} : () -> !fp32
+    %0 = tile.cion max, none, %cst, %arg0 {sink = #map0, srcs = [#map1]} : !fp32, tensor<1x784x!eltwise.fp32> -> tensor<784x!eltwise.fp32>
+    return %0 : tensor<784x!eltwise.fp32>
+  }
+}
+''').lstrip())
```

Comment:
Why don't we just trim both lhs and rhs? You can make a helper function to make it nicer.